nativelib/src/main/scala/scala/re2s/Machine.scala

```diff
@@ -0,0 +1,400 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Original Go source here:
+// http://code.google.com/p/go/source/browse/src/pkg/regexp/exec.go
+
+package scala.re2s
+
+import java.util.ArrayList
+import java.util.Arrays
+import java.util.List
+import scala.util.control.Breaks._
+
+import Inst.{Op => IOP}
+
+// A Machine matches an input string of Unicode characters against an
+// RE2 instance using a simple NFA.
+//
+// Called by RE2.doExecute.
+class Machine(re2: RE2) {
+  import Machine._
+
+  // Compiled program.
+  private val prog = re2.prog
+
+  // Two queues for runq, nextq.
+  private val q0 = new Queue(prog.numInst())
+  private val q1 = new Queue(prog.numInst())
+
+  // pool of available threads
+  // Really a stack:
+  private val pool = new ArrayList[Thread]()
+
+  // Whether a match was found.
+  private var matched: Boolean = _
+
+  // Capture information for the match.
+  private var matchcap =
+    new Array[Int](if (prog.numCap < 2) 2 else prog.numCap)
+
+  // init() reinitializes an existing Machine for re-use on a new input.
+  def init(ncap: Int): Unit = {
+    val iter = pool.iterator()
+    while (iter.hasNext()) {
+      val t = iter.next()
+      t.cap = new Array[Int](ncap)
+    }
+    this.matchcap = new Array[Int](ncap)
+  }
+
+  def submatches(): Array[Int] = {
+    if (matchcap.length == 0) {
+      return Utils.EMPTY_INTS
+    }
+    val cap = new Array[Int](matchcap.length)
+    System.arraycopy(matchcap, 0, cap, 0, matchcap.length)
+    cap
+  }
+
+  // alloc() allocates a new thread with the given instruction.
+  // It uses the free pool if possible.
+  private def alloc(inst: Inst): Thread = {
+    val n = pool.size()
+    val t = if (n > 0) pool.remove(n - 1) else new Thread(matchcap.length)
+    t.inst = inst
+    t
+  }
+
+  // free() returns t to the free pool.
+  private def free(t: Thread): Unit =
+    pool.add(t)
+
+  // match() runs the machine over the input |in| starting at |pos| with the
+  // RE2 Anchor |anchor|.
+  // It reports whether a match was found.
+  // If so, matchcap holds the submatch information.
+  def match_(in: MachineInput, _pos: Int, anchor: Int): Boolean = {
+    var pos       = _pos
+    val startCond = re2.cond
+    if (startCond == Utils.EMPTY_ALL) { // impossible
+      return false
+    }
+    if ((anchor == RE2.ANCHOR_START || anchor == RE2.ANCHOR_BOTH) &&
+        pos != 0) {
+      return false
+    }
+    matched = false
+    Arrays.fill(matchcap, -1)
+    var runq   = q0
+    var nextq  = q1
+    var r      = in.step(pos)
+    var rune   = r >> 3
+    var width  = r & 7
+    var rune1  = -1
+    var width1 = 0
+    if (r != MachineInput.EOF) {
+      r = in.step(pos + width)
+      rune1 = r >> 3
+      width1 = r & 7
+    }
+    var flag = 0 // bitmask of EMPTY_* flags
+    if (pos == 0) {
+      flag = Utils.emptyOpContext(-1, rune)
+    } else {
+      flag = in.context(pos)
+    }
+
+    breakable {
```

Comment:
I will look into this.  I had been treating the re2s code as a black box and focusing on
ensuring the output passes tests.  Looks like I have to level up and look inside ;-)

I took one run at this and will do another when my mind is clear.  The code is _ugly_
in the max. It does breaks inside of if statement blocks, with no else clause. So the 
now-known-to-be-brainless edits did on my first pass blew up and cost me half a day.
The excuse "It made sense at the time" never worked in my (extended) childhood
and it does not seem to be working now.

I need to trace the code carefully but this looks like some kind of state machine distorted
by successive language translations. It worked for the translators and I am using
their work, so I am not complaining, just trying to figure things out.

I may look at the original Go code to see the ancestor. Perhaps there is some
clarity there.   

I personally am of the school that Exceptions should be used for exceptional cases, not
normal flow control.

Reply:
Prompted by your comment, I grep`d all other re2s source and unit-test files.  The .scala
files had the one `breakable` you caught.  


There is one unit-test with a `breakable`.  It is not performance critical but it is on my
list to fix also.

Reply:
As recommended, I removed the `breakable` idiom from Machine.match_().  That is a
key piece of re2s, so I think all of re2s will benefit from the now faster code.

Travis CI is all green.

Reply:
Looks much better now! 👍

Reply:
Thank you.  Can I ask for some guidance, if you please?  

ExecTest.scala is the only other place now in the SN re2s codebase which uses
`breakable`.  There are two instances, in two logically distinct places.  At runtime this
results in, I think, six executions. The same transform of the `breakable` to a 
subroutine/method can be done but may not be economical. 

ExecTest.scala is a test suite (yes, naming convention inconsistency, jury still out on that one)
intended to be "on-demand" and not done every CI run. It is an exhaustive test which takes
tens of minutes (20? 30?  I walk away and go to lunch.)  It should be available in the repository
so that developers of SN itself can find and use it in their private directories.  It gives 
comfort when one knows that it runs successfully.

Expunge `breakable` from the code base has benefit but is saving six executions over
the span of 20 or more minutes on a test run once every human lunar landing worth
the devo time?

If yes, I will proceed when I come up for air from Unicode.   A long read for a one word answer ;-)

Reply:
It's perfectly fine to keep `breakable` in tests. My comment above only concerns non-test code.

Reply:
Understood.  Thank you for the guidance.